pkg/controller/servicebinding/secret.go

```diff
@@ -48,36 +86,39 @@ func (s *secret) createOrUpdate(payload map[string][]byte, ownerReference metav1
 
 	resourceClient := s.buildResourceClient()
 
-	logger.Debug("Attempt to create secret...")
 	existingSecret, err := s.get()
 	if err != nil {
 		if errors.IsNotFound(err) {
-			_, err := resourceClient.Create(u, metav1.CreateOptions{})
+			logger.Info("Secret not found. Creating secret")
+			secretNew, err := resourceClient.Create(u, metav1.CreateOptions{})
 			if err != nil {
 				logger.Error(err, "Error creating secret")
 				return nil, err
 			}
 			logger.Info("Secret created")
-			return u, nil
+			return secretNew, nil
 		}
 		return nil, err
 	}
-	existingSecretData, _, _ := unstructured.NestedMap(existingSecret.Object, "data")
-	payloadInterim := make(map[string]interface{})
-	for k, v := range payload {
-		payloadInterim[k] = base64.StdEncoding.EncodeToString(v)
+
+	// compare current and existing secret name
+	if u.GetName() != existingSecret.GetName() {
+		return nil, errs.New("Current and Existing Secret name not same")
 	}
-	comparisonResult := nestedMapComparison(existingSecretData, payloadInterim)
-	if comparisonResult.Success {
-		logger.Debug("Secret data is same. Skip Update")
-	} else {
-		logger.Info("Secret data is different; update secret", "Diff", comparisonResult.Diff)
-		_, err = resourceClient.Update(u, metav1.UpdateOptions{})
-		if err != nil {
-			return nil, err
-		}
+
+	// compare current and existing secret data
+	payload := make(map[string]interface{})
+	for k, v := range s.data {
+		payload[k] = base64.StdEncoding.EncodeToString(v)
 	}
-	return u, nil
+
+	existingSecretData, _, _ := unstructured.NestedMap(existingSecret.Object, "data")
+
+	if !nestedMapComparison(payload, existingSecretData).Success {
+		return nil, errs.New("Current and Existing Secret data not same")
```

Comment:
How is this error recovered? Is there a code path that fixes the content of the secret if it has been modified by the user?

Reply:
I have added the code path which updates the secret @isutton